### ISSUE TITLE
Error in inference.py

### ISSUE THREAD

**Issue by erlebach** [pymc-devs/pymc#5902]
I am looking at inference.py, trying to understand the code. Here is a snippet. It is not clear to me where `self.approx` is ever defined? Looking at the total file, `self.approx` is never set. On the other hand, there is a line, before the end of the function `__init__` and the function `_maybe_score` with the line: 
``` python 
  approx = property(lambda self:self.objective.approx)
```
This line looks strange. Perhaps this is where `self.approx` should be set? 

Here is the code snippet mentioned above: 
``` python
class Inference:
    r"""**Base class for Variational Inference**
    Communicates Operator, Approximation and Test Function to build Objective Function
    Parameters
    ----------
    op: Operator class
    approx: Approximation class or instance
    tf: TestFunction instance
    model: Model
        PyMC Model
    kwargs: kwargs passed to :class:`Operator`
    """

    def __init__(self, op, approx, tf, **kwargs):
        self.hist = np.asarray(())
        self.objective = op(approx, **kwargs)(tf)
        self.state = None

    approx = property(lambda self: self.objective.approx)

    def _maybe_score(self, score):
        returns_loss = self.objective.op.returns_loss
        if score is None:
            score = returns_loss
        elif score and not returns_loss:
            warnings.warn(
                "method `fit` got `score == True` but %s "
                "does not return loss. Ignoring `score` argument" % self.objective.op
            )
            score = False
        else:
            pass
        return score
```

**Comment by erlebach** [comment]
I retract the question. I believe that self.approx is set through `op(approx, **kwargs)(tf)`. This code is rather hard to follow!

**Comment by ricardoV94** [comment]
> This code is rather hard to follow!

Let us know if you have ideas on how to refactor the module to be more readable. 

Feel free to open a topic in our discourse for any questions or suggestions.

Closing this in the meantime.
